STYLEGUIDE.md

```diff
@@ -0,0 +1,1336 @@
+
+#### Use JMH for running benchmarks and stress tests
+In order to create benchmarks it is advised to use [JMH](http://tutorials.jenkov.com/java-performance/jmh.html). *One must run those tests on a clean controlled environment in order to obtain meaningful results*. One can use the `JunitRunner` to perform assertions on the results. Of course if you use JUnit you must make sure it is excluded from your other unit tests.
+
+    :::java
+    public class BenchmarkRunner {
+
+    @Test
+    public void launchDbBenchmarks() throws RunnerException {
+        Options opts = new OptionsBuilder()
+                .include(RocksDbBenchmark.class.getName() + ".*")
+                .mode(Mode.AverageTime)
+                .timeUnit(TimeUnit.MILLISECONDS)
+                .warmupIterations(5)
+                .forks(1)
+                .measurementIterations(10)
+                .shouldFailOnError(true)
+                .shouldDoGC(false)
+                .build();
+
+        //possible to do assertions over run results
+        Collection<RunResult> runResults = new Runner(opts).run();
+    }
+
+
+
+#### Thread.sleep()
+Sleeping is rarely warranted, especially in test code.  Sleeping is expressing an expectation that
+something else is happening while the executing thread is suspended.  This quickly leads to
+brittleness; for example if the background thread was not scheduled while you were sleeping.
+
+Sleeping in tests is also bad because it sets a firm lower bound on how fast tests can execute.
+No matter how fast the machine is, a test that sleeps for one second can never execute in less than
+one second.  Over time, this leads to very long test execution cycles.
+
+### Avoid randomness in tests
+Using random values may seem like a good idea in a test, as it allows you to cover more test cases with less code.  The problem is that you lose control over which test cases you're covering.  When you do encounter a test failure, it may be difficult to reproduce. Pseudorandom input with a fixed seed is slightly better, but in practice rarely improves test coverage.  In general it's better to use fixed input data that exercises known edge cases.
+
+## Best practices
+
+### Defensive programming
+
+#### Avoid assert
+We avoid the assert statement since it can be
+[disabled](http://docs.oracle.com/javase/7/docs/technotes/guides/language/assert.html#enable-disable) at execution time, and prefer to enforce these types of invariants at all times.
+
+*See also [preconditions](#preconditions)*
+
+#### Preconditions
+Preconditions checks are a good practice, since they serve as a well-defined barrier against bad input from callers.  As a convention, object parameters to public constructors and methods should always be checked against null, unless null is explicitly allowed. It is advised to use `Objects.requireNonNull("Informative error message", object)` to throw a preemptive `NullPointerExceptions`.
```

Comment:
`Objects.requireNonNull("Informative error message", object)`  - the order is swapped.

Reply:
Great catch!